h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostUtils.java

```diff
@@ -534,65 +537,92 @@ private static int initalizeFromChunkIds(Frame f, int[] chunks, Vec.Reader[] vec
                 // enlarge final data arrays by 2x if needed
 
                 for (int j = 0; j < di._cats; ++j) {
-                    data[currentRow][currentCol] = 1; //one-hot encoding
+                    if(colPointer == ARRAY_MAX){
+                        colPointer = 0;
+                        rowPointer++;
+                    }
+                    data[rowPointer][colPointer] = 1; //one-hot encoding
                     if (vecs[j].isNA(i)) {
-                        colIndex[currentRow][currentCol++] = di.getCategoricalId(j, Float.NaN);
+                        colIndex[rowPointer][colPointer++] = di.getCategoricalId(j, Float.NaN);
                     } else {
-                        colIndex[currentRow][currentCol++] = di.getCategoricalId(j, vecs[j].at8(i));
+                        colIndex[rowPointer][colPointer++] = di.getCategoricalId(j, vecs[j].at8(i));
                     }
                     nonZeroCount++;
                 }
 
                 for (int j = 0; j < di._nums; ++j) {
+                    if(colPointer == ARRAY_MAX){
+                        colPointer = 0;
+                        rowPointer++;
+                    }
                     float val = (float) vecs[di._cats + j].at(i);
                     if (val != 0) {
-                        data[currentRow][currentCol] = val;
-                        colIndex[currentRow][currentCol++] = di._catOffsets[di._catOffsets.length - 1] + j;
+                        data[rowPointer][colPointer] = val;
+                        colIndex[rowPointer][colPointer++] = di._catOffsets[di._catOffsets.length - 1] + j;
                         nonZeroCount++;
                     }
                 }
-
-                rowHeaders[0][++actualRows] = nonZeroCount;
```

Comment:
This was usually sufficient, as the number of rows was never larger than the first dimension of this arrray.